### PR TITLE
Fix popover toggle buttons

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Alternative to #27406.

I don't think there's a need to introduce another prop and public API to the Popover component. When `onClose` is called, we can check if the active element is the toggle button and let the toggle button handle closing instead. The Popover component should not concern itself with toggle buttons and what controls it.

Ideally this should only be implemented once with a drop down or toggleable popover component instead of needing to handle this in different places.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by stokesman** [comment]
>Ideally this should only be implemented once with a drop down or toggleable popover component instead of needing to handle this in different places.

I'd been tinkering with an approach like that and made #30397. It has test failures at the moment but I believe they may all be snags on implementation detail.

Relating to the implementation in this PR, I'd like to note that checking against `activeElement` is why the current `Dropdown` implementation has issues in some UAs (Safari and Firefox on macOS) that don't focus buttons when they are pressed. I'd explored making the `Button` component force itself to gain focus on mousedown and that can work to resolve most toggle issues but it's not the approach I used in the PR I've linked.

**Comment by ellatrix** [comment]
If the button doesn't receive focus, will the blur event still happen?

**Comment by stokesman** [comment]
> If the button doesn't receive focus, will the blur event still happen?

Yes because the button's closest focusable container gets focus.

**Comment by stokesman** [comment]
Oh, something I just remembered, buttons inside toolbars are receiving focus even in UAs that don't normally focus them (I think it's something done by ToolbarItem/reakit). This means all the toggles/popovers affected by this PR will work fine with the logic used.

**Comment by ellatrix** [comment]
@stokesman So does this test well or is there anything missing?

**Comment by stokesman** [comment]
I tested today in Safari and the toggles in `PostDate` and `ButtonEdit` are fixed but found that the one in `NavigationLinkEdit` was not. Which is actually consistent with other toggles inside toolbars in Safari. I'd previously said:
> buttons inside toolbars are receiving focus even in UAs that don't normally focus them 

But that was due to testing in Firefox and doesn't apply to Safari (though it is **supposed** to with [reakit's focus normalization](https://github.com/reakit/reakit/blob/caa81b385c0e95689a13a729635374c60956b154/packages/reakit/src/Tabbable/Tabbable.ts#L45-L51)). So now it seems odd that this even works for `PostDate` and `ButtonEdit` in Safari.

I tried a couple of structure tweaks to the toolbar in `NavigationLinkEdit` but had no luck getting it to work.

Besides all that, I know of one more toggle that wants fixing, Format Library's link component.

File: packages/block-library/src/post-date/edit.js [hunk 1]
```diff
@@ -77,6 +78,7 @@ export default function PostDateEdit( { attributes, context, setAttributes } ) {
 						<ToolbarButton
 							icon={ edit }
 							title={ __( 'Change Date' ) }
+							dateToggleButton={ toggleButtonRef }
```
Comment (stokesman) [inline packages/block-library/src/post-date/edit.js]: I happened to spot this as I ended up rebasing this and fixing the merge conflict before testing today. Changed it before testing.
```
							ref={ toggleButtonRef }
```

**Comment by stokesman** [comment]
Not sure if I jumped the gun but I tested this and it fixes what it's supposed to (tested in Safari and Chromium). Yet I did run into some trouble with the post date popover not closing when somewhere outside the block is clicked (both browers). Here's a recording:

https://user-images.githubusercontent.com/9000376/133294499-cd41d481-3589-45ca-8a12-829b019f62e1.mp4

**Comment by stokesman** [comment]
Oh yeah, yet another new popover toggle which could use this treatment is in [`DuotoneControl`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/duotone-control/index.js).

**Comment by stokesman** [comment]
I'm back again 👋 . I just made #35024 to help break this up a bit. It fixes the toggle/popover combos in `PostDateEdit` and `DuotoneControl` by replacing them with `Dropdown`. In the past, I've tried the same for the ones in the Button and Navigation Link blocks but they have requirements that `Dropdown` can't currently meet.

So if that PR looks good, between it and this one, I think we'd finally knock out all the misbehaving toggles.

**Comment by stokesman** [comment]
I assumed it'd be okay to rebase and fix conflicts here and did so.

**Pull request closed without merging**